src/fu-common.c

```diff
@@ -1413,3 +1413,73 @@ fu_common_strnsplit (const gchar *str, gsize sz,
 	}
 	return g_strsplit (str, delimiter, max_tokens);
 }
+
+/**
+ * fu_memcpy_safe:
+ * @dst: destination buffer
+ * @dst_sz: maximum size of @dst, typically `sizeof(dst)`
+ * @dst_offset: offset in bytes into @dst to copy to
+ * @src: source buffer
+ * @src_sz: maximum size of @dst, typically `sizeof(src)`
+ * @src_offset: offset in bytes into @src to copy from
+ * @n: number of bytes to copy from @src+@offset from
+ * @error: A #GError or %NULL
+ *
+ * Copies some memory using memcpy in a safe way. Providing the buffer sizes
+ * of both the destination and the source allows us to check for buffer overflow.
+ *
+ * Providing the buffer offsets also allows us to check reading past the end of
+ * the source buffer. For this reason the caller should NEVER add an offset to
+ * @src or @dst.
+ *
+ * You don't need to use this function in "obviously correct" cases, nor should
+ * you use it when performance is a concern. Only us it when you're not sure if
+ * malicious data from a device or firmware could cause memory corruption.
+ *
+ * Return value: %TRUE if the bytes were copied, %FALSE otherwise
+ **/
+gboolean
+fu_memcpy_safe (guint8 *dst, gsize dst_sz, gsize dst_offset,
```

Comment:
I'm curious; why not use `memcpy_s` from C11 instead?  It pretty much does the same thing - no?

Reply:
It's frustratingly close to being the same, but not quite. `memcpy_s()` has this prototype:

    void *dest, size_t destSize, const void *src, size_t count

So, if you do the common case of `memcpy_s(dst + FOO_DST_OFFSET, sizeof(dst), src + FOO_SRC_OFFSET, sizeof(src))` then that's insecure still. If you include the index as part of the memcpy implementation you can do the index check in one place. Using the above example, you'd think a safe `memcpy_s` would look like this: `memcpy_s(dst + FOO_DST_OFFSET, sizeof(dst) - FOO_DST_OFFSET, src + FOO_SRC_OFFSET)` but that's still wrong as you can read past the end of `@src` i.e. it prevents a crash but means you read junk and never know.

Reply:
also, as part of writing this patch I had to think about "what exactly is the size of this buffer I#m reading and writing to" quite a lot, which is what I wanted to happen.

Reply:
Gotcha, well looks good to me then.